contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/TIICallbackServlet.java

```diff
@@ -0,0 +1,146 @@
+		M_log.debug("doPost TIICallbackServlet");
+		String ipAddress = request.getRemoteAddr();
+		M_log.debug("Service request from IP=" + ipAddress);
+		
+		String allowOutcomes = ServerConfigurationService.getString(SakaiBLTIUtil.BASICLTI_OUTCOMES_ENABLED, SakaiBLTIUtil.BASICLTI_OUTCOMES_ENABLED_DEFAULT);
+		if ( ! "true".equals(allowOutcomes) ) allowOutcomes = null;
+
+		if (allowOutcomes == null ) {
+			M_log.warn("LTI Services are disabled IP=" + ipAddress);
+			response.setStatus(HttpServletResponse.SC_FORBIDDEN);
+			return;
+		}
+		
+		String contentType = request.getContentType();
+		if ( contentType != null && contentType.startsWith("application/json") ) {
+			doPostJSON(request, response);
+		} else {
+			M_log.warn("TIICallbackServlet received a not json call. Callback should have a Content-Type header of application/json.");
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+		}
+	}
+	
+	@SuppressWarnings("unchecked")
+    protected void doPostJSON(HttpServletRequest request, HttpServletResponse response) 
+            throws ServletException, IOException 
+    {
+		M_log.debug("doPostJSON TIICallbackServlet");
+		
+		IMSJSONRequest jsonRequest = new IMSJSONRequest(request);
+		if ( jsonRequest.valid ) {
+			M_log.debug(jsonRequest.getPostBody());
+		}
+		
+		String turnitinSite = ServerConfigurationService.getString("turnitin.lti.site", "!turnitin");
+		Map<String,Object> tiiData = ServletUtils.obtainGlobalTurnitinLTITool(turnitinSite);
+		if(tiiData == null){
+			M_log.error("Turnitin global LTI tool does not exist or properties are wrongly configured.");
+		}
+		String key = String.valueOf(tiiData.get(LTIService.LTI_CONSUMERKEY));
+		String secret = String.valueOf(tiiData.get(LTIService.LTI_SECRET));
+		
+		// Lets check the signature
+		if ( key == null || secret == null ) {
+			M_log.debug("doPostJSON Deployment is missing credentials");
+			response.setStatus(HttpServletResponse.SC_FORBIDDEN); 
+			doErrorJSON(request, response, jsonRequest, "Deployment is missing credentials", null);
+			return;
+		}
+
+		jsonRequest.validateRequest(key, secret, request);
+		if ( !jsonRequest.valid ) {
+			M_log.debug("doPostJSON OAuth signature failure");
+			response.setStatus(HttpServletResponse.SC_FORBIDDEN); 
+			doErrorJSON(request, response, jsonRequest, "OAuth signature failure", null);
+			return;
+		}
+
+		JSONObject json = new JSONObject(jsonRequest.getPostBody());
+		//M_log.debug(json.toString());
+		
+		String assignmentId = json.getString("resource_link_id");
+		int tiiId = json.getInt("assignmentid");
+		//ext_resource_tool_placement_url parameter can also be processed if necessary
+		try{
+			Session session = SessionManager.getCurrentSession();
+			session.setUserId("admin");
+			AssignmentContentEdit ace = AssignmentService.editAssignmentContent(assignmentId);
+			ResourcePropertiesEdit aPropertiesEdit = ace.getPropertiesEdit();
+			aPropertiesEdit.addProperty("turnitin_id", String.valueOf(tiiId));
```

Comment:
While this property is currently only set and used by the Turnitin implementation, it should be named generically. I'm using "cr_external_asn_id" locally. That way we can expose the idea of an external assignment id across implementations.